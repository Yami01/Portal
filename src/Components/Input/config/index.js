/**
 *
 * @format
 * @flow
 *
 */
export const INPUT_LINE_HEIGHT_1X = 24;
export const INPUT_PADDING_VERTICAL = 16;
