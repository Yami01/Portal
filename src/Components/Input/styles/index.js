/**
 *
 * @format
 * @flow
 *
 */

import {StyleSheet} from 'react-native';

import {
  COLOR_AIA_DARK_GREY_500,
  WHITE,
  TRANSPARENT,
  COLOR_AIA_DARK_GREY_950,
  COLOR_AIA_GREY_100,
} from '@colors';

import type {InputStylesType} from '../types';

const InputStyles: InputStylesType = StyleSheet.create({
  container: {
    paddingRight: 12,
    flexDirection: 'row',
    width: '100%',
    borderWidth: 1,
    borderRadius: 8,
    borderColor: COLOR_AIA_DARK_GREY_500,
    justifyContent: 'center',
    backgroundColor: WHITE,
    overflow: 'hidden',
  },
  disabledContainer: {
    backgroundColor: COLOR_AIA_GREY_100,
  },
  inputContainer: {
    overflow: 'hidden',
    flex: 1,
  },
  input: {
    fontSize: 17,
    lineHeight: 24,
    paddingLeft: 0,
    backgroundColor: TRANSPARENT,
    overflow: 'hidden',
    paddingBottom: 5,
    textAlign: 'auto',
  },
  rightComponentContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    minWidth: 30,
  },
  rightComponentText: {
    fontSize: 12,
    lineHeight: 17,
    color: COLOR_AIA_DARK_GREY_950,
  },
});

export default InputStyles;
