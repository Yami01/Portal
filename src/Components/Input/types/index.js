/**
 *
 * @format
 * @flow
 *
 */
import type {
  ViewStyleProp,
  TextStyleProp,
} from 'react-native/Libraries/StyleSheet/StyleSheet'

export type KeyboardType =
  | 'default'
  | 'number-pad'
  | 'decimal-pad'
  | 'numeric'
  | 'email-address'
  | 'phone-pad'
export type PropsType = {
  testID?: string,
  keyboardType?: KeyboardType,
  secureTextEntry?: boolean,
  label?: string,
  editable?: boolean,
  disabled?: boolean,
  pointerEvents?: 'auto' | 'none' | 'box-none',
  value: ?string,
  decimalPlaces?: number,
  onChangeText?: string => void,
  onBlur?: () => void,
  rightComponentType?: 'icon' | 'text',
  rightText?: string,
  iconGroup?: string,
  iconName?: string,
  iconWidth?: number,
  iconHeight?: number,
  iconFillColor?: string,
  maxLength?: number,
  placeholder?: string,
  textInputContainerStyle?: ViewStyleProp,
  textInputStyle?: TextStyleProp,
}

export type InputStylesType = {
  inputContainer: ViewStyleProp,
  container: ViewStyleProp,
  disabledContainer: ViewStyleProp,
  input: TextStyleProp,
  rightComponentContainer: ViewStyleProp,
  rightComponentText: TextStyleProp,
}
