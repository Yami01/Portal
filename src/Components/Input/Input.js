/**
 *
 * Input
 * @format
 * @flow
 *
 */

import React, { memo, useMemo, useCallback } from 'react'
import type { Node, AbstractComponent } from 'react'
import { View } from 'react-native'
import { TextInput } from 'react-native-paper'
import { INPUT_LINE_HEIGHT_1X, INPUT_PADDING_VERTICAL } from './config'
import InputStyles from './styles'
import type { PropsType } from './types'
import Text from '@/Components/Text'
import { Colors } from '@/Theme/Variables'
import Icon from '@/Components/Icon'
import { useFontScale } from '@/Hooks'

const Input: AbstractComponent<PropsType> = memo((props: PropsType): Node => {
  const {
    testID,
    keyboardType,
    secureTextEntry,
    label,
    editable,
    disabled,
    pointerEvents,
    value,
    decimalPlaces,
    onChangeText,
    onBlur: onBlurProp,
    rightComponentType,
    rightText,
    iconGroup,
    iconName,
    iconWidth,
    iconHeight,
    iconFillColor,
    maxLength,
    placeholder,
    textInputContainerStyle,
    textInputStyle,
  } = props

  const fontScale = useFontScale()

  const inputHeight = useMemo((): number => {
    return INPUT_LINE_HEIGHT_1X * fontScale + 2 * INPUT_PADDING_VERTICAL
  }, [fontScale])

  const textNode = useMemo((): Node => {
    return (
      <View>
        <Text style={InputStyles.rightComponentText}>{rightText}</Text>
      </View>
    )
  }, [rightText])

  const iconNode = useMemo((): Node => {
    return (
      <Icon
        group={iconGroup || ''}
        name={iconName || ''}
        fillColor={iconFillColor}
        width={iconWidth}
        height={iconHeight}
      />
    )
  }, [iconGroup, iconName, iconWidth, iconHeight, iconFillColor])

  const rightComponent = useMemo((): Node => {
    if (!rightComponentType) {
      return null
    }

    return rightComponentType === 'text' ? textNode : iconNode
  }, [rightComponentType, textNode, iconNode])

  // Ensure no params pass to TextInput onBlur
  const onBlur = useCallback(() => {
    onBlurProp?.()
  }, [onBlurProp])

  const validateInputValue = useCallback(
    (inputValue: string) => {
      if (!onChangeText) {
        return
      }

      if (decimalPlaces && decimalPlaces > 0) {
        const pattern = new RegExp(`^(\\d*\\.{0,1}\\d{0,${decimalPlaces}}$)`)
        const validated = inputValue.match(pattern)

        if (validated) {
          onChangeText(inputValue)
        }
      } else {
        onChangeText(inputValue)
      }
    },
    [decimalPlaces, onChangeText],
  )

  return (
    <View
      style={[
        InputStyles.container,
        disabled && InputStyles.disabledContainer,
        textInputContainerStyle,
      ]}
    >
      <View style={[InputStyles.inputContainer, { height: inputHeight }]}>
        <TextInput
          {...testID}
          style={[InputStyles.input, { height: inputHeight }, textInputStyle]}
          value={value}
          disabled={disabled}
          editable={editable}
          pointerEvents={pointerEvents}
          keyboardType={keyboardType}
          secureTextEntry={secureTextEntry}
          onChangeText={validateInputValue}
          onBlur={onBlur}
          blurOnSubmit
          label={label}
          underlineColor={Colors.transparent}
          placeholder={placeholder}
          maxLength={maxLength}
          theme={{
            colors: {
              primary: Colors.gray_950,
              text: Colors.gray_750,
            },
          }}
        />
      </View>
      {rightComponentType && (
        <View style={InputStyles.rightComponentContainer}>
          {rightComponent}
        </View>
      )}
    </View>
  )
})

export default Input
