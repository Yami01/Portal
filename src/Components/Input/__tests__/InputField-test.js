/**
 *
 * Input Test
 * @format
 * @flow
 *
 */

import React from 'react';
import renderer from 'react-test-renderer';
import {render, fireEvent} from '@testing-library/react-native';

import {COLOR_AIA_DARK_GREY_650} from '@colors';

import Input from '../Input';

describe('Input', () => {
  const props = {
    accessibilityLabel: 'Input',
    accessibilityHint: 'Please Input',
    label: 'Textinput',
    value: '',
    onChangeText: () => {
      jest.fn();
    },
  };

  const iconProps = {
    rightComponentType: 'icon',
    iconGroup: 'general',
    iconName: 'timer',
    iconFillColor: COLOR_AIA_DARK_GREY_650,
    iconAccessibilityLabel: 'Icon',
    iconAccessibilityHint: 'Icon Hint',
  };

  const textProps = {
    rightComponentType: 'text',
    rightText: 'test',
    rightTextAccessibilityLabel: 'Text label',
    rightTextAccessibilityHint: 'Text Hint',
  };

  test('Input Render Properly', () => {
    const tree = renderer.create(<Input {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Input & Icon Accessibility', () => {
    const {queryByA11yLabel, queryByA11yHint} = render(<Input {...props} {...iconProps} />);
    expect(queryByA11yLabel(props.accessibilityLabel)).not.toBeNull();
    expect(queryByA11yHint(props.accessibilityHint)).not.toBeNull();
    expect(queryByA11yLabel(iconProps.iconAccessibilityLabel)).not.toBeNull();
    expect(queryByA11yHint(iconProps.iconAccessibilityHint)).not.toBeNull();
  });

  test('Input RHS Text Component Accessibility', () => {
    const {queryByA11yLabel, queryByA11yHint, queryByText} = render(
      <Input {...props} {...textProps} />
    );
    expect(queryByA11yLabel(textProps.rightTextAccessibilityLabel)).not.toBeNull();
    expect(queryByA11yHint(textProps.rightTextAccessibilityHint)).not.toBeNull();
    expect(queryByText(textProps.rightText)).not.toBeNull();
  });

  test('onChangeText handler is called', () => {
    const onChangeTextMock = jest.fn();
    const mockValue = 'Test';
    const {getByDisplayValue} = render(
      <Input {...props} value={mockValue} onChangeText={onChangeTextMock} />
    );

    const input = getByDisplayValue(mockValue);
    fireEvent.changeText(input, 'Test');
    expect(onChangeTextMock).toHaveBeenCalled();
  });

  test('value is set by value prop', () => {
    const value = 'Hello, world.';
    const {queryByDisplayValue} = render(<Input {...props} value={value} />);

    const input = queryByDisplayValue(value);
    expect(input).not.toBeNull();
  });

  test("onChangeText is called with the value if the value's decimal places is less than or equal to to decimalPlaces prop", () => {
    const testID = 'inputTestID';
    const mockOnChangeText = jest.fn();
    const mockValue = '100.00';
    const decimalPlaces = 2;

    const {getByTestId} = render(
      <Input
        {...props}
        testID={testID}
        onChangeText={mockOnChangeText}
        decimalPlaces={decimalPlaces}
      />
    );

    const input = getByTestId(testID);

    fireEvent.changeText(input, mockValue);

    expect(mockOnChangeText).toHaveBeenCalledWith(mockValue);
  });

  test("onChangeText is not called if the value's decimal places is more than decimalPlaces prop", () => {
    const testID = 'inputTestID';
    const mockOnChangeText = jest.fn();
    const mockValue = '100.000';
    const decimalPlaces = 2;

    const {getByTestId} = render(
      <Input
        {...props}
        testID={testID}
        onChangeText={mockOnChangeText}
        decimalPlaces={decimalPlaces}
      />
    );

    const input = getByTestId(testID);

    fireEvent.changeText(input, mockValue);

    expect(mockOnChangeText).not.toHaveBeenCalled();
  });
});
