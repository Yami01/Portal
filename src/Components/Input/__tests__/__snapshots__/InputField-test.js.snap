// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Input Render Properly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "borderColor": "#868b8c",
        "borderRadius": 8,
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "overflow": "hidden",
        "paddingRight": 12,
        "width": "100%",
      },
      undefined,
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "overflow": "hidden",
        },
        Object {
          "height": 80,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(231, 231, 231)",
            "borderTopLeftRadius": 4,
            "borderTopRightRadius": 4,
          },
          Object {
            "backgroundColor": "rgba(0,0,0,0)",
            "lineHeight": 24,
            "overflow": "hidden",
            "paddingBottom": 5,
            "paddingLeft": 0,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "rgba(0,0,0,0)",
            "bottom": 0,
            "height": 2,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "transform": Array [
              Object {
                "scaleY": 0.5,
              },
            ],
          }
        }
      />
      <View
        style={
          Array [
            Object {
              "paddingBottom": 0,
              "paddingTop": 0,
            },
            Object {
              "minHeight": 80,
            },
          ]
        }
      >
        <View
          pointerEvents="none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <Text
            numberOfLines={1}
            onLayout={[Function]}
            style={
              Object {
                "color": "#5c646c",
                "fontFamily": "System",
                "fontSize": 17,
                "fontWeight": undefined,
                "left": 0,
                "opacity": 0,
                "paddingLeft": 12,
                "paddingRight": 12,
                "position": "absolute",
                "textAlign": "left",
                "top": 40,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                ],
                "writingDirection": "ltr",
              }
            }
          >
            Textinput
          </Text>
          <Text
            numberOfLines={1}
            style={
              Object {
                "color": "rgba(0, 0, 0, 0.54)",
                "fontFamily": "System",
                "fontSize": 17,
                "fontWeight": undefined,
                "left": 0,
                "opacity": 0,
                "paddingLeft": 12,
                "paddingRight": 12,
                "position": "absolute",
                "textAlign": "left",
                "top": 40,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                ],
                "writingDirection": "ltr",
              }
            }
          >
            Textinput
          </Text>
        </View>
        <TextInput
          accessibilityHint="Please Input"
          accessibilityLabel="Input"
          allowFontScaling={true}
          blurOnSubmit={true}
          editable={true}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder=""
          placeholderTextColor="rgba(0, 0, 0, 0.54)"
          rejectResponderTermination={true}
          selectionColor="#5c646c"
          style={
            Array [
              Object {
                "flexGrow": 1,
                "margin": 0,
                "zIndex": 1,
              },
              Object {
                "paddingLeft": 12,
                "paddingRight": 12,
              },
              Object {
                "height": 80,
              },
              Object {
                "paddingBottom": 4,
                "paddingTop": 24,
              },
              Object {
                "color": "#333d47",
                "fontFamily": "System",
                "fontSize": 17,
                "fontWeight": undefined,
                "textAlign": "auto",
                "textAlignVertical": "center",
              },
              false,
              Array [
                Object {},
              ],
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
  </View>
</View>
`;
